File: frontend/src/player/engine.ts
/**
 * Deterministic story player engine (Story Runtime Semantics v1), TypeScript port.
 *
 * This mirrors the Python reference engine
 * (`src/cyo_adventure/player/engine.py`) exactly so the player and validator
 * never disagree. The shared player-trace conformance corpus
 * (`schema/conformance/player_traces.json`) is run by both implementations.
 *
 * Transition order on every choice: evaluate condition -> apply choice effects
 * -> set current_node -> apply target on_enter effects (once:true first entry).
 * The engine is pure: choose() returns a new ReadingState and never mutates input.
 */

import { evaluate } from './evaluator'
import type {
  Choice,
  Effect,
  ReadingState,
  SavedBookmark,
  Storybook,
  StoryNode,
  VarState,
} from './types'

function nodeIndex(story: Storybook): Map<string, StoryNode> {
  return new Map(story.nodes.map((node) => [node.id, node]))
}

function intBounds(story: Storybook): Map<string, [number | null, number | null]> {
  const bounds = new Map<string, [number | null, number | null]>()
  for (const v of story.variables) {
    if (v.type === 'int') {
      bounds.set(v.name, [v.min ?? null, v.max ?? null])
    }
  }
  return bounds
}

function clamp(
  bounds: Map<string, [number | null, number | null]>,
  name: string,
  value: number
): number {
  const [low, high] = bounds.get(name) ?? [null, null]
  if (low !== null && value < low) return low
  if (high !== null && value > high) return high
  return value
}

function applyEffect(
  varState: VarState,
  effect: Effect,
  bounds: Map<string, [number | null, number | null]>
): void {
  if (effect.op === 'set') {
    const value = effect.value ?? 0
    // Clamp a numeric set to the variable's bounds, like inc/dec, so the TS and
    // Python engines agree and a story cannot seed an out-of-range value.
    varState[effect.var] = typeof value === 'number' ? clamp(bounds, effect.var, value) : value
    return
  }
  const current = varState[effect.var]
  const base = typeof current === 'number' ? current : 0
  const delta = typeof effect.value === 'number' ? effect.value : 0
  const updated = effect.op === 'inc' ? base + delta : base - delta
  varState[effect.var] = clamp(bounds, effect.var, updated)
}

function enterNode(
  story: Storybook,
  state: ReadingState,
  nodeId: string,
  firstEntry: boolean,
  bounds: Map<string, [number | null, number | null]>
): void {
  if (!state.visit_set.includes(nodeId)) {
    state.visit_set.push(nodeId)
  }
  const node = nodeIndex(story).get(nodeId)
  // Mirror the Python engine: entering an unknown node id is an error, not a
  // silent no-op, so a dangling choice target fails loudly in both runtimes.
  if (!node) {
    throw new Error(`node '${nodeId}' does not exist in the story`)
  }
  for (const effect of node.on_enter ?? []) {
    if (effect.once && !firstEntry) continue
    applyEffect(state.var_state, effect, bounds)
  }
}

// #ASSUME: data-integrity: enterNode trusts that nodeId exists in the story.
// A dangling target (choice pointing to a non-existent node) throws immediately,
// matching the Python engine and the Layer-1 L1-2 reference-integrity check.
// #VERIFY: the validator rejects stories with dangling targets before they reach
// the reader, so this throw is a belt-and-suspenders guard, not a normal path.

/** Begin a new read at start_node with initial variable values. */
export function start(story: Storybook): ReadingState {
  const varState: VarState = {}
  for (const v of story.variables) {
    varState[v.name] = v.initial
  }
  const state: ReadingState = {
    current_node: story.start_node,
    var_state: varState,
    path: [story.start_node],
    visit_set: [],
    version: story.version,
    state_revision: 0,
    save_slots: {},
  }
  enterNode(story, state, story.start_node, true, intBounds(story))
  return state
}

// #CRITICAL: data-integrity: a continuation state cannot be reproduced by
// replaying choices from start_node, so continuation saves MUST NOT carry a
// choice_path (the server would replay-from-start and reject them; see
// api/reading.py's note that choice_path may become required). The server's
// structural floor (player/replay.py::_check_structure) is what admits these
// saves, so this function must uphold its exact invariants: every declared
// variable present, values correctly typed and in-bounds (clamped below),
// current_node === path[path.length - 1], all node ids known.
// #VERIFY: engine.test.ts "startContinuation" describe block; if choice_path
// ever becomes required server-side, the server needs a continuation-aware
// replay mode first.
/** Begin a continuation read at a declared entry node, seeding name-matched
 * carried variables (WS-G decision G3). Wrong-typed carried values are
 * skipped (the declared initial stands); carried ints are clamped to the
 * variable's declared bounds. */
export function startContinuation(
  story: Storybook,
  entryNode: string | null,
  carriedVarState?: VarState
): ReadingState {
  const bounds = intBounds(story)
  const varState: VarState = {}
  for (const v of story.variables) {
    varState[v.name] = v.initial
    const carried = carriedVarState?.[v.name]
    if (carried === undefined) continue
    if (v.type === 'bool' && typeof carried === 'boolean') {
      varState[v.name] = carried
    } else if (v.type === 'int' && typeof carried === 'number' && Number.isInteger(carried)) {
      varState[v.name] = clamp(bounds, v.name, carried)
    }
  }
  const nodeId =
    entryNode !== null && story.nodes.some((n) => n.id === entryNode) ? entryNode : story.start_node
  const state: ReadingState = {
    current_node: nodeId,
    var_state: varState,
    path: [nodeId],
    visit_set: [],
    version: story.version,
    state_revision: 0,
    save_slots: {},
  }
  enterNode(story, state, nodeId, true, bounds)
  return state
}

/** Choices visible at the current node (false-condition choices are hidden). */
export function visibleChoices(story: Storybook, state: ReadingState): Choice[] {
  const node = nodeIndex(story).get(state.current_node)
  if (!node) return []
  return node.choices.filter((c) => c.condition == null || evaluate(c.condition, state.var_state))
}

/** Whether the current node is an ending. */
export function isEnding(story: Storybook, state: ReadingState): boolean {
  return nodeIndex(story).get(state.current_node)?.is_ending ?? false
}

/** The stable ending id of the current node, if it is an ending. */
export function currentEndingId(story: Storybook, state: ReadingState): string | null {
  const node = nodeIndex(story).get(state.current_node)
  return node?.is_ending ? (node.ending?.id ?? null) : null
}

// #CRITICAL: timing: choose() transition order (condition check -> choice effects
// -> set current_node -> on_enter effects) MUST stay in sync with the Python
// reference engine (src/cyo_adventure/player/engine.py). Divergence causes the
// Layer-2 validator and the runtime to disagree on reachable states.
// #VERIFY: shared player_traces.json conformance corpus is run by both engines.

/** Apply a choice and return the resulting reading state (input is not mutated). */
export function choose(story: Storybook, state: ReadingState, choiceId: string): ReadingState {
  if (isEnding(story, state)) {
    throw new Error(`cannot choose from ending node '${state.current_node}'`)
  }
  const node = nodeIndex(story).get(state.current_node)
  const choice = node?.choices.find((c) => c.id === choiceId)
  if (!choice) {
    throw new Error(`choice '${choiceId}' does not exist on the current node`)
  }
  if (!(choice.condition == null || evaluate(choice.condition, state.var_state))) {
    throw new Error(`choice '${choiceId}' is not visible in the current state`)
  }
  // #ASSUME: data-integrity: intBounds is rebuilt per choose() call; bounds are
  // not cached across calls. A cached bounds map would go stale if the Storybook
  // object is replaced (e.g. on a story update) without clearing the cache.
  // #VERIFY: choose() receives a fresh story reference on each call from the reader.
  const bounds = intBounds(story)
  const next: ReadingState = {
    current_node: state.current_node,
    var_state: { ...state.var_state },
    path: [...state.path],
    visit_set: [...state.visit_set],
    version: state.version,
    state_revision: state.state_revision,
    save_slots: { ...state.save_slots },
  }
  for (const effect of choice.effects ?? []) {
    applyEffect(next.var_state, effect, bounds)
  }
  next.current_node = choice.target
  const firstEntry = !next.visit_set.includes(choice.target)
  enterNode(story, next, choice.target, firstEntry, bounds)
  next.path.push(choice.target)
  return next
}

// ---------------------------------------------------------------------------
// Go back one page (kid mis-tap recovery). A frontend-only affordance built on
// replay: the previous state is recomputed by replaying the recorded node path
// from the start through this same deterministic engine, never by reversing
// effects, so a post-back state is exactly the state of a shorter read. No new
// state semantic is introduced, so nothing needs mirroring in the Python engine.
// ---------------------------------------------------------------------------

// #EDGE: timing: the path replay backtracks across same-target sibling choices
// (a node may offer two choices to the same target with different effects,
// like the lantern story's take/ignore pair), which is exponential in the
// pathological case. Real stories are short and near-deterministic; this
// budget bounds the search and fails closed (no Go back) if exhausted.
// #VERIFY: engine.test.ts replays the ambiguous lantern ignore-lantern branch.
const MAX_REPLAY_STEPS = 5000

function sameVarState(a: VarState, b: VarState): boolean {
  const aKeys = Object.keys(a)
  return aKeys.length === Object.keys(b).length && aKeys.every((key) => a[key] === b[key])
}

/** Order-independent id-set equality, mirroring how the backend replay gate
 * compares visit_set (player/replay.py::_check_replay uses set equality). */
function sameIdSet(a: string[], b: string[]): boolean {
  const aSet = new Set(a)
  const bSet = new Set(b)
  return aSet.size === bSet.size && [...aSet].every((id) => bSet.has(id))
}

interface ReplayBudget {
  remaining: number
}

/** Depth-first reconstruction of live.path: at each step try every visible
 * choice whose target is the next recorded node, and accept only a branch
 * whose end state reproduces the live state. Appends to states in place;
 * states[i] is the state after i choices when the search succeeds. */
function searchPathReplay(
  story: Storybook,
  live: ReadingState,
  states: ReadingState[],
  budget: ReplayBudget
): boolean {
  const depth = states.length - 1
  const current = states[depth]
  if (depth === live.path.length - 1) {
    // current_node and path match by construction; the variables and visit set
    // must match too, so an unfaithful reconstruction (a different same-target
    // sibling than the one actually taken) is rejected, not rewritten into
    // the child's history.
    return (
      sameVarState(current.var_state, live.var_state) &&
      sameIdSet(current.visit_set, live.visit_set)
    )
  }
  const targetId = live.path[depth + 1]
  for (const candidate of visibleChoices(story, current)) {
    if (candidate.target !== targetId) continue
    if (budget.remaining <= 0) return false
    budget.remaining -= 1
    let next: ReadingState
    try {
      next = choose(story, current, candidate.id)
    } catch {
      // A dangling target throws inside choose(); treat the branch as dead and
      // fail closed rather than crash the reader on a corrupt story.
      continue
    }
    states.push(next)
    if (searchPathReplay(story, live, states, budget)) return true
    states.pop()
  }
  return false
}

// #VERIFY: engine.test.ts "offers Go back on a seeded read", "still fails
// closed when the recorded path does not start where it should", and
// "fails closed for a continuation state (path does not begin at
// start_node)".
/** Replay live.path from the read's own start (the seeded start when a seed
 * is given, story.start_node otherwise), returning the state after each
 * recorded step (result[i] is the state after i choices), or null when no
 * replay of the recorded path reproduces the live state. */
export function replayRecordedPath(
  story: Storybook,
  live: ReadingState,
  seed?: VarState
): ReadingState[] | null {
  // The recorded path must begin where a read with THIS seed begins. The
  // seeded start supplies the carried variables the read actually began
  // with; replaying from declared initials made the terminal var_state
  // comparison in searchPathReplay fail for every seeded read, so Go back
  // was disabled. entryNode is null here, so initial.current_node is still
  // story.start_node in every case: the node comparison below is unchanged,
  // and the fail-closed guarantee it enforces is unchanged with it.
  let initial: ReadingState
  try {
    initial = seed === undefined ? start(story) : startContinuation(story, null, seed)
  } catch {
    // A dangling start node: fail closed, same as the dead-branch case above.
    return null
  }
  if (live.path.length === 0 || live.path[0] !== initial.current_node) return null
  const states = [initial]
  return searchPathReplay(story, live, states, { remaining: MAX_REPLAY_STEPS }) ? states : null
}

// #ASSUME: data-integrity: a state saved after Go back is indistinguishable
// from having simply made fewer choices, so the existing save path needs no
// change. Verified against src/cyo_adventure/api/reading.py::put_reading_state
// (revision-based optimistic concurrency: 409 only on version/state_revision
// mismatch; nothing requires path to grow between saves) and
// player/replay.py::_check_structure (known node ids, current_node ===
// path[path.length - 1], complete in-bounds var_state), all of which a
// replayed shorter state satisfies by construction; choice_path is optional
// and the frontend does not send it.
// #VERIFY: ReaderPage stamps state_revision from its own revisionRef before
// each PUT, so the revision carried over below never fights the server
// counter; tests/unit/test_replay.py pins the structural floor server-side.
/** The reading state as if the child had made every recorded choice except
 * the last one, recomputed via replay (never by reversing effects); null when
 * there is nothing to undo or the recorded path cannot be faithfully replayed
 * from the read's own start (the seeded start when a seed is given). The
 * input is not mutated. */
export function back(story: Storybook, state: ReadingState, seed?: VarState): ReadingState | null {
  if (state.path.length <= 1) return null
  const states = replayRecordedPath(story, state, seed)
  if (states === null) return null
  const previous = states[states.length - 2]
  return {
    current_node: previous.current_node,
    var_state: { ...previous.var_state },
    path: [...previous.path],
    visit_set: [...previous.visit_set],
    version: previous.version,
    // Rewind only what choices produced: the server-revision counter and save
    // slots are owned outside the choice history and carry over unchanged.
    state_revision: state.state_revision,
    save_slots: { ...state.save_slots },
  }
}

/** Whether Go back is available: at least one recorded choice, and the
 * recorded path is faithfully replayable from the read's own start. */
export function canGoBack(story: Storybook, state: ReadingState, seed?: VarState): boolean {
  return back(story, state, seed) !== null
}

// ---------------------------------------------------------------------------
// Bookmarks (save-slot feature, distinct from the Go back undo above): a
// named snapshot of a reading position a reader chooses to save, so they can
// keep reading past it and return later. Stored in the existing
// `save_slots` field, which the backend has always persisted, synced across
// devices, and byte-capped opaquely (api/schemas.py); nothing here changes
// the wire contract, only what this client chooses to put in that bag.
// ---------------------------------------------------------------------------

const MAX_BOOKMARKS = 10

/** Runtime shape guard: a save_slots value written by a future feature (or a
 * stale format from before this one existed) must never be treated as a
 * bookmark and crash the bookmarks UI; it is silently excluded instead. */
function isSavedBookmark(value: unknown): value is SavedBookmark {
  if (value === null || typeof value !== 'object') return false
  const v = value as Record<string, unknown>
  return (
    typeof v.current_node === 'string' &&
    typeof v.var_state === 'object' &&
    v.var_state !== null &&
    Array.isArray(v.visit_set) &&
    v.visit_set.every((id) => typeof id === 'string') &&
    Array.isArray(v.path) &&
    v.path.every((id) => typeof id === 'string') &&
    typeof v.label === 'string' &&
    typeof v.saved_at === 'string'
  )
}

/** Every bookmark currently saved, newest first. Any `save_slots` entry that
 * does not look like a bookmark (see `isSavedBookmark`) is silently
 * excluded, never thrown on. */
export function listBookmarks(state: ReadingState): Array<{ id: string; bookmark: SavedBookmark }> {
  return Object.entries(state.save_slots)
    .filter((entry): entry is [string, SavedBookmark] => isSavedBookmark(entry[1]))
    .map(([id, bookmark]) => ({ id, bookmark }))
    .sort((a, b) => b.bookmark.saved_at.localeCompare(a.bookmark.saved_at))
}

/** Whether another bookmark can be saved right now.
 *
 * #ASSUME: data-integrity: 10 is a client-side UX ceiling, not derived from
 * the backend's 64,000-byte save_slots budget (a single bookmark snapshot,
 * dominated by var_state and a short path, is far smaller than that); it
 * exists so a reader cannot turn the bookmarks list into an unusable wall of
 * entries, not to protect the byte cap. #VERIFY: BookmarksPanel.test.tsx
 * asserts Save is disabled at the ceiling.
 */
export function canSaveBookmark(state: ReadingState): boolean {
  return listBookmarks(state).length < MAX_BOOKMARKS
}

/** Save the CURRENT position as a new bookmark under a fresh slot id.
 *
 * Pure and immutable like every other function in this module: returns a
 * new ReadingState, never mutates `state`. `slotId`/`savedAt` are injected
 * (crypto.randomUUID() / new Date().toISOString() at the real call site,
 * player/machine.ts) rather than generated here, so this function stays as
 * deterministically testable as `choose()`/`back()`.
 */
export function saveBookmark(
  state: ReadingState,
  slotId: string,
  label: string,
  savedAt: string
): ReadingState {
  const bookmark: SavedBookmark = {
    current_node: state.current_node,
    var_state: { ...state.var_state },
    visit_set: [...state.visit_set],
    path: [...state.path],
    label,
    saved_at: savedAt,
  }
  return {
    ...state,
    save_slots: { ...state.save_slots, [slotId]: bookmark },
  }
}

/** Remove a bookmark. A missing/already-removed slotId is a no-op (matches
 * `Record` delete semantics), not an error: a doubled delete tap (a slow
 * network response racing a second tap) must never throw. */
export function deleteBookmark(state: ReadingState, slotId: string): ReadingState {
  const nextSlots = { ...state.save_slots }
  delete nextSlots[slotId]
  return { ...state, save_slots: nextSlots }
}

/** Make a saved bookmark the new LIVE position (current_node/var_state/
 * visit_set/path), keeping save_slots (including this same bookmark)
 * unchanged, so "jump to bookmark" does not also delete it.
 *
 * Returns null, never throws, when slotId does not resolve to a bookmark
 * (deleted by another device since the list was rendered, or a corrupt
 * save_slots value) -- the same fail-closed contract as `back()` above, for
 * the same reason: this is driven by a tap on a list the UI already
 * rendered from a possibly-stale snapshot.
 */
export function loadBookmark(state: ReadingState, slotId: string): ReadingState | null {
  const raw = state.save_slots[slotId]
  if (!isSavedBookmark(raw)) return null
  return {
    current_node: raw.current_node,
    var_state: { ...raw.var_state },
    visit_set: [...raw.visit_set],
    path: [...raw.path],
    version: state.version,
    state_revision: state.state_revision,
    save_slots: { ...state.save_slots },
  }
}

File: src/cyo_adventure/player/engine.py
"""Deterministic story player engine (Story Runtime Semantics v1).

This is the reference implementation of the canonical execution model in
``docs/planning/runtime-semantics.md``. It is pure: no I/O, no async, no shared
mutable state. The Layer-2 validator walk (Phase 2) and the TypeScript player
must reproduce its behaviour exactly, and the cross-implementation conformance
fixtures exist to prove they do.

Transition order on every choice (section 1):

1. Evaluate the choice condition against the current ``var_state``.
2. Apply the choice effects.
3. Set ``current_node`` to the choice target.
4. Apply the target node's ``on_enter`` effects (``once: true`` first-entry only).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from cyo_adventure.core.exceptions import BusinessLogicError
from cyo_adventure.storybook.evaluator import evaluate

if TYPE_CHECKING:
    from cyo_adventure.storybook.evaluator import VarState
    from cyo_adventure.storybook.models import Choice, Effect, Node, Storybook

from cyo_adventure.player.state import ReadingState


class StoryEngine:
    """Plays a single parsed :class:`Storybook` deterministically."""

    def __init__(self, story: Storybook) -> None:
        """Index a story for traversal.

        Args:
            story: The parsed, schema-valid Storybook to play.
        """
        self._story = story
        self._nodes: dict[str, Node] = {node.id: node for node in story.nodes}
        self._bounds: dict[str, tuple[int | None, int | None]] = {
            var.name: (var.min, var.max)
            for var in story.variables
            if var.type.value == "int"
        }

    def start(self) -> ReadingState:
        """Begin a new read at ``start_node`` with initial variable values.

        Entering the start node applies its ``on_enter`` effects (first entry).

        Returns:
            ReadingState: The initial reading state pinned to the story version.
        """
        return self.start_continuation(None)

    def start_continuation(
        self, carried: VarState | None, entry_node: str | None = None
    ) -> ReadingState:
        """Begin a read at ``entry_node``, seeding name-matched carried variables.

        Mirrors the client's ``startContinuation``
        (``frontend/src/player/engine.ts``) exactly, including the order of
        operations: carried values are merged into the declared initials
        **before** the entry node's ``on_enter`` effects run. Seeding afterwards
        would let a carried value overwrite an effect the entry node applied,
        which the client does not do, and player/validator divergence here is a
        conformance failure under runtime-semantics.md section 1.

        The entry-node parameter closes a real divergence rather than adding a
        feature (`UW-C296`). The client has taken an ``entryNode`` since WS-G and
        ``reader/ContinueSeries.tsx`` passes the receiving book's
        ``series_entry_node``; this method did not, so SR-9 validated a reader
        path that begins at ``start_node`` while the reader begins somewhere
        else. A comment here previously justified that by asserting the two are
        equal for every book, which is false on committed content: both
        brass-lantern books declare ``start_node: n_open`` against
        ``series_entry_node: n_start``.

        Fallback matches the client exactly: an entry node of ``None``, or one no
        node in the story carries, starts at ``start_node``.

        Carry rules (WS-G G3, identical to the client): a carried value is used
        only when its name matches a declared variable and its type matches;
        a wrong-typed value is skipped so the declared initial stands; a carried
        int is clamped into the receiving variable's declared bounds.

        Args:
            carried: Carried variable values from a predecessor book, or
                ``None`` for an ordinary read.
            entry_node: The node id to begin at, or ``None`` for ``start_node``.
                An id absent from the story falls back to ``start_node``.

        Returns:
            ReadingState: The initial reading state pinned to the story version.
        """
        var_state: VarState = {var.name: var.initial for var in self._story.variables}
        if carried:
            for var in self._story.variables:
                value = carried.get(var.name)
                if value is None:
                    continue
                if var.type.value == "bool" and isinstance(value, bool):
                    var_state[var.name] = value
                elif (
                    var.type.value == "int"
                    and isinstance(value, int)
                    and not isinstance(value, bool)
                ):
                    var_state[var.name] = self._clamp(var.name, value)
        # #CRITICAL: data integrity: the node entered here must be the node the
        # reader enters, or every state-seeded rule validates a path nobody
        # walks. The existence guard mirrors the client's
        # `story.nodes.some(n => n.id === entryNode)` rather than trusting the
        # declaration, so a stale or misspelled entry node degrades to the start
        # node on both sides instead of diverging.
        # #VERIFY: test_series.py::test_sr9_walks_the_declared_series_entry_node
        # pins that a non-start entry changes SR-9's carried reachability, and
        # test_player_conformance.py covers the client parity.
        node_id = (
            entry_node
            if entry_node is not None
            and any(node.id == entry_node for node in self._story.nodes)
            else self._story.start_node
        )
        state = ReadingState(
            current_node=node_id,
            var_state=var_state,
            path=[node_id],
            visit_set=set(),
            version=self._story.version,
        )
        self._enter_node(state, node_id, first_entry=True)
        return state

    def visible_choices(self, state: ReadingState) -> list[Choice]:
        """Return the choices visible at the current node for this state.

        A choice with a condition that evaluates to false is hidden (section 4).

        Args:
            state: The current reading state.

        Returns:
            list[Choice]: The choices the reader may select right now.
        """
        node = self._node(state.current_node)
        return [c for c in node.choices if self._is_visible(c, state.var_state)]

    def is_ending(self, state: ReadingState) -> bool:
        """Report whether the current node is an ending.

        Args:
            state: The current reading state.

        Returns:
            bool: ``True`` if the current node is an ending node.
        """
        return self._node(state.current_node).is_ending

    def current_ending_id(self, state: ReadingState) -> str | None:
        """Return the stable ending id of the current node, if it is an ending.

        Args:
            state: The current reading state.

        Returns:
            str | None: The ``ending.id`` or ``None`` if not at an ending.
        """
        node = self._node(state.current_node)
        return node.ending.id if node.is_ending and node.ending is not None else None

    def choose(self, state: ReadingState, choice_id: str) -> ReadingState:
        """Apply a choice and return the resulting reading state.

        The returned state is a fresh object; the input ``state`` is not mutated,
        so a caller may keep the prior state (for example, for a save slot).

        Args:
            state: The current reading state.
            choice_id: The id of the choice to select.

        Returns:
            ReadingState: The state after the transition.

        Raises:
            BusinessLogicError: If the current node is an ending, the choice id
                is unknown at the current node, or the choice is not visible.
        """
        if self.is_ending(state):
            msg = f"cannot choose from ending node '{state.current_node}'"
            raise BusinessLogicError(msg)
        node = self._node(state.current_node)
        choice = next((c for c in node.choices if c.id == choice_id), None)
        if choice is None:
            msg = f"choice '{choice_id}' does not exist on node '{node.id}'"
            raise BusinessLogicError(msg)
        if not self._is_visible(choice, state.var_state):
            msg = f"choice '{choice_id}' is not visible in the current state"
            raise BusinessLogicError(msg)
        # Transition order (runtime-semantics section 1): the input state is
        # snapshotted so the engine stays pure and replays are deterministic.
        nxt = self._clone(state)
        for effect in choice.effects:
            self._apply_effect(nxt.var_state, effect)
        target = choice.target
        nxt.current_node = target
        first_entry = target not in nxt.visit_set
        self._enter_node(nxt, target, first_entry=first_entry)
        nxt.path.append(target)
        return nxt

    def _enter_node(
        self, state: ReadingState, node_id: str, *, first_entry: bool
    ) -> None:
        """Mark a node visited and apply its ``on_enter`` effects.

        A ``once: true`` effect applies only on the first entry to the node
        (runtime-semantics section 2).

        Args:
            state: The state to mutate in place.
            node_id: The node being entered.
            first_entry: Whether this is the first time the node is entered.
        """
        state.visit_set.add(node_id)
        node = self._node(node_id)
        for effect in node.on_enter:
            if effect.once and not first_entry:
                continue
            self._apply_effect(state.var_state, effect)

    def _apply_effect(self, var_state: VarState, effect: Effect) -> None:
        """Apply a single effect to a variable state, honoring int bounds.

        Args:
            var_state: The variable state to mutate in place.
            effect: The effect to apply.
        """
        op = effect.op.value
        if op == "set":
            value = effect.value if effect.value is not None else 0
            # A set to a bounded int variable is clamped like inc/dec so a story
            # cannot seed an out-of-range value that conditions later compare
            # against (runtime-semantics section 3); non-int values pass through.
            if isinstance(value, int) and not isinstance(value, bool):
                value = self._clamp(effect.var, value)
            var_state[effect.var] = value
            return
        current = var_state.get(effect.var, 0)
        delta = effect.value if isinstance(effect.value, int) else 0
        # #ASSUME: data integrity: schema guarantees `current` is int for inc/dec
        # targets (L1-6); clamping is a defensive fallback only (semantics 3).
        # #VERIFY: validator rejects bound-exceeding transitions before publish.
        base = current if isinstance(current, int) else 0
        updated = base + delta if op == "inc" else base - delta
        var_state[effect.var] = self._clamp(effect.var, updated)

    def _clamp(self, var_name: str, value: int) -> int:
        """Clamp a value to the variable's declared bounds, if any.

        Args:
            var_name: The variable being updated.
            value: The proposed new value.

        Returns:
            int: The value bounded by the declared ``min``/``max``.
        """
        low, high = self._bounds.get(var_name, (None, None))
        if low is not None and value < low:
            return low
        if high is not None and value > high:
            return high
        return value

    def _is_visible(self, choice: Choice, var_state: VarState) -> bool:
        """Return whether a choice is visible under the current variable state.

        Args:
            choice: The choice to test.
            var_state: The current variable state.

        Returns:
            bool: ``True`` if the choice has no condition or it evaluates true.
        """
        if choice.condition is None:
            return True
        return evaluate(choice.condition, var_state)

    def _node(self, node_id: str) -> Node:
        """Look up a node by id.

        Args:
            node_id: The node id to resolve.

        Returns:
            Node: The node.

        Raises:
            BusinessLogicError: If the node id is not in the story.
        """
        node = self._nodes.get(node_id)
        if node is None:
            msg = f"node '{node_id}' does not exist in story '{self._story.id}'"
            raise BusinessLogicError(msg)
        return node

    @staticmethod
    def _clone(state: ReadingState) -> ReadingState:
        """Return a deep-ish copy of a reading state for a pure transition.

        Args:
            state: The state to copy.

        Returns:
            ReadingState: A copy whose mutable containers are independent.
        """
        return ReadingState(
            current_node=state.current_node,
            var_state=dict(state.var_state),
            path=list(state.path),
            visit_set=set(state.visit_set),
            version=state.version,
            state_revision=state.state_revision,
            # Copy each snapshot, not just the dict, so a cloned timeline cannot
            # mutate a save slot shared with the original (purity guarantee).
            save_slots={name: snap.copy() for name, snap in state.save_slots.items()},
        )
